File: server/src/main/scala/producer/message_generator/data_generators/int64/Int64Generator.scala
package producer.message_generator.data_generators.int64

import com.tools.teal.pulsar.ui.producer.v1.producer as pb
import producer.message_generator.json_generator.JsonGenerator

case class Int64Generator(
    generator: Long | JsonGenerator
)

object Int64Generator:
    def fromPb(v: pb.Int64Generator): Int64Generator =
        val generator = v.generator match
            case pb.Int64Generator.Generator.GeneratorFixedInt64(v) => v
            case pb.Int64Generator.Generator.GeneratorFromJson(v) => JsonGenerator.fromPb(v)
        Int64Generator(
            generator = generator
        )

    def toPb(v: Int64Generator): pb.Int64Generator =
        val generator = v.generator match
            case v: Long => pb.Int64Generator.Generator.GeneratorFixedInt64(v)
            case v: JsonGenerator => pb.Int64Generator.Generator.GeneratorFromJson(JsonGenerator.toPb(v))

        pb.Int64Generator(
           generator = generator
        )

File: ui/components/ui/ProducerSession/message-generator/data-generators/int64/int64-generator.ts
import * as pb from '../../../../../../grpc-web/tools/teal/pulsar/ui/producer/v1/producer_pb';
import { JsonGenerator, jsonGeneratorFromPb } from '../json-generator/json-generator';

export type Int64Generator = {
  type: 'int64-generator',
  generator: { type: 'fixed-int64', n: number } | { type: 'from-json', jsonGenerator: JsonGenerator };
};

export function int64GeneratorFromPb(v: pb.Int64Generator): Int64Generator {
  let generator: Int64Generator['generator'];
  switch (v.getGeneratorCase()) {
    case pb.Int64Generator.GeneratorCase.GENERATOR_FIXED_INT64:
      generator = { type: 'fixed-int64', n: v.getGeneratorFixedInt64() };
      break;
    case pb.Int64Generator.GeneratorCase.GENERATOR_FROM_JSON:
      generator = { type: 'from-json', jsonGenerator: jsonGeneratorFromPb(v.getGeneratorFromJson()!) };
      break;
    default:
      throw new Error(`Unknown int64 generator case: ${v.getGeneratorCase()}`);
  }

  return { type: 'int64-generator', generator };
}
